lib/Backend/arm64/LegalizeMD.cpp

```diff
@@ -804,29 +804,75 @@ void LegalizeMD::LegalizeLdLabel(IR::Instr * instr, IR::Opnd * opnd)
     }
 }
 
-bool LegalizeMD::LegalizeDirectBranch(IR::BranchInstr *branchInstr, uint32 branchOffset)
+namespace
+{
+    IR::LabelInstr* TryInsertBranchIsland(IR::Instr* instr, IR::LabelInstr* target, int limit, bool forward)
+    {
+        int i = MachMaxInstrSize;
+        IR::Instr* branchInstr = nullptr;
+
+        if (forward)
+        {
+            for (IR::Instr* next = instr->m_next; i < limit && next != nullptr; next = next->m_next)
+            {
+                if (next->IsBranchInstr() && next->AsBranchInstr()->IsUnconditional())
+                {
+                    branchInstr = next;
+                    break;
+                }
+                i += MachMaxInstrSize;
+            }
+        }
+        else
+        {
+            for (IR::Instr* prev = instr->m_prev; i < limit && prev != nullptr; prev = prev->m_prev)
+            {
+                if (prev->IsBranchInstr() && prev->AsBranchInstr()->IsUnconditional())
+                {
+                    branchInstr = prev;
+                    break;
+                }
+                i += MachMaxInstrSize;
+            }
+        }
+
+        if (branchInstr == nullptr)
+        {
+            return nullptr;
+        }
+
+        IR::LabelInstr* islandLabel = IR::LabelInstr::New(Js::OpCode::Label, instr->m_func, false);
+        branchInstr->InsertAfter(islandLabel);
+
+        IR::BranchInstr* targetBranch = IR::BranchInstr::New(Js::OpCode::B, target, branchInstr->m_func);
+        islandLabel->InsertAfter(targetBranch);
+
+        return islandLabel;
+    }
+}
+
+bool LegalizeMD::LegalizeDirectBranch(IR::BranchInstr *branchInstr, uintptr_t branchOffset)
 {
     Assert(branchInstr->IsBranchInstr());
 
-    uint32 labelOffset = (uint32)branchInstr->GetTarget()->GetOffset();
+    uintptr_t labelOffset = branchInstr->GetTarget()->GetOffset();
     Assert(labelOffset); //Label offset must be set.
 
-    int32 offset = labelOffset - branchOffset;
+    intptr_t offset = labelOffset - branchOffset;
     //We should never run out of 26 bits which corresponds to +-64MB of code size.
```

Comment:
This "offset" is the byte offset. However, all the sizes you are checking against are actually word (4-byte) offsets. You should either divide this by 4, or increase the range checking below.

Reply:
For example, TBZ/TBNZ have a 14-bit displacement field, so it can reference +/-8k words, which is actually +/-32k bytes.

Reply:
Does this apply to the IS_CONST_INT21 for ADR relocs as well? From the encoder I don't think it does, but want to confirm.

Reply:
No, there is no implicit shift in ADR. But definitely all the branches have them.